### PR TITLE
Make UAuthority store binary IP addresses, update tests, and address bugs found along the way

### PR DESCRIPTION
**Pull request by gregmedd**
This PR addresses, among other things, #67 - `UAuthority` was storing human readable string addresses instead of the "byte" representation required by the spec. It is based on the pending changes from #68, and will be blocked until that PR is merged.

Along the way, I stumbled into several other bugs and deviations from spec. For full details of changes made, I recommend referring to the commit messages for my commits. The short summary of changes is:

- Switch to binary IP addresses
- Move `IpAddress` class from `uri/serializer` to `uri/tools`
- Add unit tests for `IpAddress` class
- Revise and extend `MicroUriSerializerTest.cpp` for better coverage
- Address issues found in the URI module with updated tests

NB: Because this PR is more extensive than initially planned, the commits have been staged such that the changes can be viewed in incremental steps, if needed. Each commit can build and run tests.

### CODE REVIEW

**Comment by PLeVasseur** [comment]
Hmmm. Is there a way to have it so that on this PR we can review the range of commits you're going to add? Like... hmm. Seems to me that'd make it a lot easier to review.

**Comment by gregmedd** [comment]
> Hmmm. Is there a way to have it so that on this PR we can review the range of commits you're going to add? Like... hmm. Seems to me that'd make it a lot easier to review.

That's why it's in draft state right now. Either I make a PR into the 1.5.7 branch, which makes my PR easier to review but the 1.5.7 PR (#74) more complicated, or this is held until the 1.5.7 PR is merged. I lean towards the latter solution as it aims to keep PRs smaller and focused on specific changes.

I could _try_ to rebase this branch onto main get it committed first, but since the 1.5.7 PR contains so many formatting changes (e.g. removing `using namespace ...`) across so many files, that is likely to further complicate things.

**Review by stevenhartley — APPROVED** [review]
Approving so it can be merged ONCE another C++ expert has reviewed and approved the change.

File: include/up-cpp/uri/tools/IpAddress.h
Comment (gregmedd) [reply include/up-cpp/uri/tools/IpAddress.h:33]: No. This is a forward declaration with no definition. This is a C/C++ism for _"this type will be defined elsewhere, but I'm going to make a passing reference to it here"_ and the compiler won't be bothered by it. It avoids `#include`'ing a file here that isn't otherwise needed.

I have compiled and tested this code, after all :grin:

**Review by debruce — APPROVED** [review]
Looks good to me.

File: include/up-cpp/uri/builder/BuildUAuthority.h
Comment (umeirov) [inline include/up-cpp/uri/builder/BuildUAuthority.h:137]: consider adding comments to explain the purpose of this method and any constraints related to the ID

File: include/up-cpp/uri/builder/BuildUAuthority.h
Comment (umeirov) [inline include/up-cpp/uri/builder/BuildUAuthority.h:46]: Consider documenting the overall purpose of the BuildUAuthority class and its intended usage

Comment (gregmedd) [reply include/up-cpp/uri/builder/BuildUAuthority.h:46]: Will address in a separate PR as that is a pre-existing component. Issue #85 has been filed to track this request.

File: include/up-cpp/uri/builder/BuildUAuthority.h
Comment (gregmedd) [reply include/up-cpp/uri/builder/BuildUAuthority.h:137]: #85 filed to track this request.

File: src/uri/IpAddress.cpp
Comment (umeirov) [inline src/uri/IpAddress.cpp:1]: This feedback doesn’t pertain to this specific file but rather addresses our overall error handling strategy. Let’s establish a consistent approach for managing errors. The current implementation lacks proper propagation to higher layers. Instead of silently proceeding with control flow and merely logging error messages, we should prioritize robust error handling across the entire code.

**Review by umeirov — APPROVED** [review]

File: include/up-cpp/uri/builder/BuildUAuthority.h
Comment (gregmedd) [reply include/up-cpp/uri/builder/BuildUAuthority.h:48]: This was pre-existing code written by another author. I agree that we should make a pass through to clean up issues like this throughout the code base. However, it probably should not be a part of this review.

Comment (gregmedd) [reply include/up-cpp/uri/builder/BuildUAuthority.h:48]: Filed as #87

File: include/up-cpp/uri/builder/BuildUUri.h
Comment (gregmedd) [reply include/up-cpp/uri/builder/BuildUUri.h:73]: It likely is not. However, this is the same method used elsewhere in the code to propagate error states. Without this, there would be no way to distinguish between an accidental attempt to set an empty entity and an entity that was never set.

Additionally, there is nothing that forces client code to use this builder interface for UUri objects, and therefore nothing stopping that code from setting an empty entity anyway. All downstream consumers of UUri would need to be aware of this possibility regardless.

In the short term, I think this trade-off is _probably_ the best course of action. We should follow up with a plan for error propagation (see discussion in #79) and address it at that time.

Comment (gregmedd) [reply include/up-cpp/uri/builder/BuildUUri.h:73]: Files as #88 so we don't lose track of it.

File: include/up-cpp/uri/builder/BuildUUri.h
Comment (gregmedd) [reply include/up-cpp/uri/builder/BuildUUri.h:87]: Filed as #88 so we don't lose track of it.

File: src/uri/IpAddress.cpp
Comment (gregmedd) [reply src/uri/IpAddress.cpp:1]: We have an open issue to discuss error propagation. Feel free to participate in #79

**Pull request merged**
